### DISCUSSION TITLE
Question: Custom error messages with variables?

### DISCUSSION THREAD

**Discussion by bulicmatko** [colinhacks/zod#1909] [Q&A] [answered]
Is there a way to set custom error messages with variables?
Here's an example to better explain my question:

 ```typescript
// An example with react hook
// useIntl comes from `react-intl`

function useValidationSchema() {
  const intl = useIntl()

  return string()
    .max(64, {
      message: intl.formatMessage({
        id: "s3BfNQ",
        defaultMessage: "Max length is 64",
      }),
    })
}
```

The problem with this example is the duplication of the max value `64` in the validator and in the translation.

Ideally, I would want something like this:

 ```typescript
// An example with react hook
// useIntl comes from `react-intl`

function useValidationSchema() {
  const intl = useIntl()

  return string()
    .max(64, {
      message: ({ max }) =>
        intl.formatMessage(
          {
            id: "s3BfNQ",
            defaultMessage: "Max length is {max}",
          },
          { max }
        ),
    })
}
```

This way, if the max value changes, there's no need to update all translations.
This is the way `yup` does it.

I checked the docs and opened issues and could not find anything like this.

PS: Big Thanks to the author and all the contributors for this great lib. Amazing work!

**Comment by JacobWeisenburger** [comment]
There are a couple of ways that I can think of to solve this.

Do these work for you?

```ts
const max = 64
z.string().max( max, `Max length is ${ max }` )
```

https://github.com/JacobWeisenburger/zod_utilz#makeerrormap
```ts
import { zu } from 'zod_utilz'

z.string( {
    errorMap: zu.makeErrorMap( {
        too_big: ( { maximum } ) => `Max length is ${ maximum }`,
    } )
} ).max( 64 )
```

**Reply by bulicmatko** [reply]
Hey @JacobWeisenburger, thanks for your reply.

The first example would work, but I feel it's not something that would be very manageable for bigger forms.
For example, one would have to define a constant for each value used as a validator option.
Let's say we have `firstName` and `lastName` and each has `min` and `max` validators. This would require four constants to be defined.

The second example could also work, but it requires a separate lib for something that feels like it should be a part of `zod` itself.

It would be nice if the value passed to the validator (min, max, length, ...) could be used directly there and then in the error message. Without the need for (over)extraction of constants.

Thanks for the examples :)

**Comment by JacobWeisenburger** [comment, marked as answer]
You could do it with just zod. But it's a bit harder to work with in my opinion.

```ts
z.string( {
    errorMap: ( issue, ctx ) => {
        if ( issue.code === z.ZodIssueCode.too_big )
            return { message: `Max length is ${ issue.maximum }` }
        return { message: ctx.defaultError }
    }
} ).max( 64 )
```

I'd recommend using [`zu.makeErrorMap`](https://github.com/JacobWeisenburger/zod_utilz#makeerrormap) to make this easier. Yes it will require another library, but it's a very small one. `npm install zod_utilz`  is a very easy command. 😉

```ts
import { zu } from 'zod_utilz'

z.string( {
    errorMap: zu.makeErrorMap( {
        too_big: ( { maximum } ) => `Max length is ${ maximum }`,
    } )
} ).max( 64 )
```

**Reply by bulicmatko** [reply]
You are right. The second example with the `zod_utilz` lib looks simpler and more to the point.
Thank you for giving a bit more insight into how this can be done in a few different ways. I'll definitely check `zod_utilz` as a solution to the problem.

That said, I would still like to hear/read more opinions and/or examples from the community regarding this.
Maybe there are more devs that prefer a solution similar to how `yup` does it.
I think this would make migration from `yup` to `zod` easier for more developers.

Thanks :)

PS: I like the idea of extracting error messages with `makeErrorMap` in `zod_utils`.  Nice ;)

**Reply by JacobWeisenburger** [reply]
do you have an example of how `yup` does it?

**Reply by bulicmatko** [reply]
In `yup` you can pass the error message as a second argument, the same as in `zod`.
The difference is that in `yup` it can be a `string` or a `function` that returns a `string`.
Here's an example:

```typescript
// Passing a string (same as zod)
string().max(64, "Max: 64");

// Passing a function
string().max(64, ({ max }) => `Max: ${max}`)
string().min(2, ({ min }) => `Min: ${min}`)
string().length(64, ({ length}) => `Length: ${length}`)
```

**Reply by JacobWeisenburger** [reply]
That's a really cool idea for a new feature. Would you like me to make this as a new issue?

**Reply by bulicmatko** [reply]
I think it would be a nice addition to this already great library! :) Thanks, Jacob!

**Reply by JacobWeisenburger** [reply]
https://github.com/colinhacks/zod/issues/1912
